diff --git a/vm/qemu/qemu.go b/vm/qemu/qemu.go
index 5f945ece..3dc2f0c0 100644
--- a/vm/qemu/qemu.go
+++ b/vm/qemu/qemu.go
@@ -68,7 +68,7 @@ type archConfig struct {
 var archConfigs = map[string]archConfig{
 	"linux/amd64": {
 		Qemu:     "qemu-system-x86_64",
-		QemuArgs: "-enable-kvm -usb -usbdevice mouse -usbdevice tablet -soundhw all",
+		QemuArgs: "-enable-kvm -usb -usbdevice mouse -usbdevice tablet ",
 	},
 	"linux/386": {
 		Qemu: "qemu-system-i386",

